src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,269 @@
+  
+  this->client_cb = cb;
+  this->cb->set_client_cb(cb);
+  aio = rgw::make_throttle(window_size, y);
+
+  uint64_t chunk_len = g_conf()->rgw_get_obj_max_req_size;
+  uint64_t num_parts = (len%chunk_len) == 0 ? len/chunk_len : (len/chunk_len) + 1;
+  uint64_t part_num = 0;
+  uint64_t cost = 0, len_to_read = 0;
+
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << "chunk_len " << chunk_len << " num_parts " << num_parts << dendl;
+  do {
+    if (part_num == (num_parts - 1)) {
+      len_to_read = len;
+      cost = len;
+    } else {
+      len_to_read = chunk_len;
+      cost = chunk_len;
+    }
+    uint64_t id = ofs;
+    uint64_t read_ofs = 0;
+    std::string oid_in_cache = source->get_bucket()->get_marker() + "_" + oid + "_" + std::to_string(ofs) + "_" + std::to_string(len_to_read) + "_" + std::to_string(part_num);
+    rgw_raw_obj r_obj;
+    r_obj.oid = oid_in_cache;
+    ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << dendl;
+    if (filter->get_d3n_cache()->get(oid_in_cache, len_to_read)) {
+      // Read From Cache
+      auto completed = aio->get(r_obj, rgw::Aio::d3n_cache_op(dpp, y, read_ofs, len_to_read, filter->get_d3n_cache()->cache_location), cost, id);
+      auto r = flush(dpp, std::move(completed));
+      if (r < 0) {
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to flush, r= " << r << dendl;
+        return r;
+      }
+    } else {
+      auto r = drain(dpp);
+      if (r < 0) {
+        ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to drain, r= " << r << dendl;
+        return r;
+      }
+      break;
+    }
+    if (part_num == (num_parts - 1)) {
+      return drain(dpp);
+    } else {
+      ofs += chunk_len;
+    }
+    part_num += 1;
+    len -= chunk_len;
+  } while(part_num < num_parts);
+
+  ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Fetching object from backend store" << dendl;
+  this->cb->set_part_num(part_num);
+  this->cb->set_ofs(ofs);
+  auto r = next->iterate(dpp, ofs, end, this->cb, y);
+  if (r < 0) {
+    ldpp_dout(dpp, 20) << "D3NFilterObject::iterate:: " << __func__ << "(): Error: failed to fetch object from backend store, r= " << r << dendl;
+    return r;
+  }
+  return this->cb->flush_last_part();
+}
+
+int D3NFilterObject::D3NFilterReadOp::D3NFilterGetCB::flush_last_part()
+{
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
+
+int D3NFilterObject::D3NFilterReadOp::D3NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  if (!last_part && bl.length() <= g_conf()->rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+    if (r < 0) {
+      return r;
+    }
+  }
+
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
```

Comment:
@cbodley : does rgw know about multipart chunk size. I couldnt see it anywhere, so I have used the logic of accumulating data in rgw_get_obj_max_req_size sized buffers and then writing the buffer to the cache.

Reply:
> @cbodley : does rgw know about multipart chunk size

no, the client can decide what part size to use for each upload, ie `s3cmd --multipart-chunk-size-mb=`

if we're only implementing the read cache here, do we need to care whether or not it's stored as multipart?

> multipart_chunk_size not taken into account - looking into that.

what issue were you seeing that needed to be fixed there?

Reply:
We dont really need to care about multipart here, just that the code deals with bufferlist copying around to make buffer sizes of rgw_get_obj_max_req_size, if rgw knew the multipart_chunk_size then handle_data would simply write data to the cache (without all the bufferlist copying and splicing), and reading would take into account multipart_chunk_size

The issue that I saw was that - while writing to the cache, the buffer was being written as is, in case of multipart uploads, if the shadow objects of a part can be less than rgw_get_obj_max_req_size, that was making the reading part incorrect (that assumed all the chunks to be of size rgw_get_obj_max_req_size, except for the last part). In order to fix this I implemented the logic of accumulating data into rgw_get_obj_max_req_size chunks before writing it out to the cache. This works for all kinds of objects regardless of them being multipart as you pointed out.

Reply:
okay, that makes sense. so the logic around `num_parts` and `part_num` in the d3n filter aren't related to multipart, they're just numbering the chunks

> bucketmarker_oid_offset_chunksize_partnum

could `bucketmarker_oid_partnum` by itself work? it seems like we could calculate the offset from that
```c++
offset = partnum * rgw_get_obj_max_req_size;
```

Reply:
yes it could, i was just making it clear for someone who would look at the cache contents to see what offset to what length is there in each part. Do you have a strong  opinion on not using the offset in the parts stored in the datacache? 

yes, num_parts and part_num are just numbering the chunks here.

Reply:
> Do you have a strong opinion on not using the offset in the parts stored in the datacache?

no, that's fine